extensions/panache/hibernate-reactive-panache/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/PanacheQuery.java

```diff
@@ -0,0 +1,264 @@
+package io.quarkus.hibernate.reactive.panache;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import javax.persistence.LockModeType;
+import javax.persistence.NoResultException;
+import javax.persistence.NonUniqueResultException;
+
+import org.hibernate.Session;
+import org.hibernate.annotations.FilterDef;
+
+import io.quarkus.panache.common.Page;
+import io.quarkus.panache.common.Parameters;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+
+/**
+ * <p>
+ * Interface representing an entity query, which abstracts the use of paging, getting the number of results, and
+ * operating on {@link List} or {@link Stream}.
+ * </p>
+ * <p>
+ * Instances of this interface cannot mutate the query itself or its parameters: only paging information can be
+ * modified, and instances of this interface can be reused to obtain multiple pages of results.
+ * </p>
+ *
+ * @author Stéphane Épardaud
+ * @param <Entity> The entity type being queried
+ */
+public interface PanacheQuery<Entity> {
```

Comment:
I really feel like `Optional` doesn't belong where we deal with `Uni` which have their own way to return empty results. I don't think we should add them.

Reply:
Maybe ask some light from the munity expert, I'm OK if we don't provide them (I'll deprecate them in MongoDB then) as soon as we explain somewhere in the doc that we don't provide them because of Mutiny or someone may ask to add them